src/ACL.php

```diff
@@ -16,33 +20,60 @@ class ACL
     public $auth;
 
     /**
-     * Given a model, this will apply some restrictions on it.
+     * Returns AccessRules model for logged in user and in model scope.
+     *
+     * @param Model $model
      *
-     * @param \atk4\data\Persistence $p
-     * @param \atk4\data\Model       $m
+     * @return \atk4\login\Model\AccessRule
      */
-    public function applyRestrictions(\atk4\data\Persistence $p, \atk4\data\Model $m)
+    public function getRules(Model $model)
     {
-        // Extend this method
-        // if($m instanceof Model\User && !$this->auth->user['is_admin']) {
-        //      $m->getField('is_admin')->read_only = true;
-        // }
-    }
+        /** @var \atk4\login\Model\User*/
+        $user = $this->auth->user;
 
-    /*
-    protected $permissions = [];
+        if (!$user->loaded()) {
+            // user is not logged in - let's force him to do so. Alternative is to throw exception, but that's ugly.
+            $this->auth->check();
+            //throw new Exception('User should be logged in!');
+        }
+
+        return $user->ref('AccessRules')->addCondition('model', get_class($model));
+    }
 
     /**
-     * Gather permissions of currently logged in users for faster access
+     * Given a model, this will apply some restrictions on it.
+     *
+     * Extend this method if you wish.
+     *
+     * @param Persistence $p
+     * @param Model       $m
      */
-    /*
-    public function cachePermissions()
+    public function applyRestrictions(Persistence $p, Model $m)
     {
-        $this->permissions = [
-            //'admin' => $this->auth->user->loaded() && $this->auth->user['is_admin'] // Imants: disabled this to not fail miserably
-        ];
+        $rules = $this->getRules($m);
+
+        foreach ($rules as $junk) {
+
+            // set visible and editable fields
+            foreach ($m->getFields() as $name => $field) {
+                $field['ui']['visible'] = $rules['all_visible'] || (array_search($name, $rules['visible_fields']) !== false);
+                $field['ui']['editable'] = $rules['all_editable'] || (array_search($name, $rules['editable_fields']) !== false);
+            }
+
+            // remove not allowed actions
+            if (!$rules['all_actions'] && $rules['actions']) {
+                $actions_to_remove = array_diff(array_keys($m->getActions()), $rules['actions']);
+                foreach ($actions_to_remove as $action) {
+                    $m->_removeFromCollection($action, 'actions');
+                }
+            }
+
+            // add conditions on model
+            if ($rules['conditions']) {
+                $m->addCondition($rules['conditions']);
```

Comment:
yeah, but need to move into separate method , as we might have ability to reference current user fields in conditions, but nothing that could cause security problems.